### PR TITLE
[IMP] l10n_it_edi: hook to manage unsupported documents types

### PR DESCRIPTION
**Pull request by valeriobelcastro**
The need for this change is to extend the document types to manage. So far the DOCUMENT TYPE code TD10-TD11-TD12-TD16-TD17-TD18-TD19-TD20-TD21-TD22-TD23-TD24 are unsupported. With this PR we have the ability to add new controls and behaviors.
we manage the message, the function is executed and at the end we integrate the invoice with the additional data.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by scigghia — APPROVED** [review]

**Review by sergiocorato — APPROVED** [review]

**Review by marcelofrare — APPROVED** [review]

**Comment by tsb-odoo** [comment]
@jco-odoo can you review this one please ?

**Comment by jco-odoo** [comment]
@valeriobelcastro What do you think about a bit broader refactoring like in https://github.com/odoo/odoo/pull/92159

**Comment by OpenCode** [comment]
> @valeriobelcastro What do you think about a bit broader refactoring like in #92159

Hi @jco-odoo I think the best scenario is the use of both the approaches. We need to intercept the document types to exclude (or eventually include) new type from the list. In your PR the document types list is still hardocodes and can't be changed/manipulated.

I propose to use this PR and then you can apply your on top of this.

**Comment by OpenCode** [comment]
> I think you could easily inherit from the more general method:
> 
> ```
> if document_type in ['TD04']:
>     return 'in_invoice', False
> return super()...
> ```

Don't focus your attention on what we need but try to see the big picture.
F.E. The function in your PR print a logger information. So, if a module call the super and then executes code to manage TD01 or TD07, the logger will show a wrong information but if we exclude the code from the list, this can't happen and the code is cleaner. In this case when keep the stack of calls and everyone lived happily ever after :D

**Comment by lordkrandel** [comment]
I'd rather have something on the line of...
```py
    def _l10n_it_document_type_mapping(self):
        return {
            'TD01': {'move_type': 'in_invoice', 'simplified': False},
            'TD04': {'move_type': 'in_refund',  'simplified': False},
            'TD07': {'move_type': 'in_invoice', 'simplified': True},
            'TD08': {'move_type': 'in_refund',  'simplified': True},
            'TD09': {'move_type': 'in_invoice', 'simplified': True},   # Simplified debit note
        }

    def _import_fattura_pa(self, tree, invoice):
        elements = tree.xpath('//DatiGeneraliDocumento/TipoDocumento')
        document_type = elements and elements[0].text
        mapped = self._l10n_it_document_type_mapping().get(document_type, None)
        if not mapped:
            _logger.info('Document type %s not recognized or managed. Invoice type is set by default.', elements[0].text)
            mapped = {'move_type': 'in_invoice', 'simplified': False}
        move_type, simplified = mapped['move_type'], mapped['simplified']
```

**Comment by OpenCode** [comment]
We are limiting the possibilities to only two values: move_type and simplified. In a different scenario, a module may have different needs and may need more data. Why limit the possibilities?

**Comment by lordkrandel** [comment]
If you have more data to put there (other than move_type and simplified), you will have more logic behind it to use that information that isn't in `_import_fattura_pa` and you will have to rewrite the function anyway.

Do you have any practical use case and an example of what you have in mind? 

Returning a dictionary of dictionaries already enables a lot of expansion.
If you don't like the part that reads the dictionary, maybe this fragment below can be more open.

```py
        mapped = self._l10n_it_document_type_mapping().get(document_type, {})
        if not mapped:
            _logger.info('Document type %s not recognized or managed. Invoice type is set by default.', elements[0].text)
        move_type = mapped.get('move_type', 'in_invoice')
        simplified = mapped.get('simplified', False)
```

**Comment by OpenCode** [comment]
Yeah. I like your propose and the propose in the @jco-odoo 's PR, too. From my pov, every step in "heritability" is a good step

File: addons/l10n_it_edi/models/account_edi_format.py [hunk 1]
```diff
@@ -671,3 +671,6 @@ def _import_fattura_pa(self, tree, invoice):
             invoices += new_invoice
 
         return invoices
+
+    def get_unsupported_document_types(self):
```
[the remark this answers is not in the document]
Comment (eLBati) [reply addons/l10n_it_edi/models/account_edi_format.py:675]: But `['TD01', 'TD07']` are **supported** types

[the remark this answers is not in the document]
Comment (valeriobelcastro) [reply addons/l10n_it_edi/models/account_edi_format.py:675]: this method should return the list of those document types for which we do not set a different move_type than the default one (which is in_invoice)... Actually you are right, _get_unsupported_document_types_ can be misleading, what do you think about something like: _get_default_move_type_document_types_?

**Pull request closed without merging**